File: generated_tasks/graphkit__6363ce2ca293/tests/hidden/base_repo/src/components.ts
import { Graph } from './graph';
import { DiGraph } from './digraph';
import { NodeId, ComponentResult } from './types';

export function connectedComponents(g: Graph): ComponentResult {
  const visited = new Set<NodeId>(); const components: NodeId[][] = []; const componentOf = new Map<NodeId, number>();
  for (const id of g.nodeIds()) {
    if (visited.has(id)) continue;
    const comp: NodeId[] = []; const stack = [id];
    while (stack.length > 0) { const u = stack.pop()!; if (visited.has(u)) continue; visited.add(u); comp.push(u); for (const v of g.neighbors(u)) if (!visited.has(v)) stack.push(v); }
    const idx = components.length; components.push(comp); for (const n of comp) componentOf.set(n, idx);
  }
  return { components, componentOf, count: components.length, largestComponent: components.reduce((a, b) => a.length >= b.length ? a : b, []) };
}

export function stronglyConnectedComponents(g: DiGraph): ComponentResult {
  const sccs = g.stronglyConnectedComponents(); const componentOf = new Map<NodeId, number>();
  sccs.forEach((c, i) => c.forEach(id => componentOf.set(id, i)));
  return { components: sccs, componentOf, count: sccs.length, largestComponent: sccs.reduce((a, b) => a.length >= b.length ? a : b, []) };
}

export function articulationPoints(g: Graph): NodeId[] {
  const visited = new Set<NodeId>(); const disc = new Map<NodeId, number>(); const low = new Map<NodeId, number>();
  const parent = new Map<NodeId, NodeId | null>(); const ap = new Set<NodeId>(); let timer = 0;
  const dfs = (u: NodeId): void => {
    visited.add(u); disc.set(u, timer); low.set(u, timer); timer++; let ch = 0;
    for (const v of g.neighbors(u)) {
      if (!visited.has(v)) {
        ch++; parent.set(v, u); dfs(v); low.set(u, Math.min(low.get(u)!, low.get(v)!));
        const isRoot = parent.get(u) === null || parent.get(u) === undefined;
        if (isRoot && ch > 1) ap.add(u);
        else if (!isRoot && low.get(v)! >= disc.get(u)!) ap.add(u);
      } else if (v !== parent.get(u)) { low.set(u, Math.min(low.get(u)!, disc.get(v)!)); }
    }
  };
  for (const id of g.nodeIds()) if (!visited.has(id)) { parent.set(id, null); dfs(id); }
  return [...ap];
}

export function bridges(g: Graph): [NodeId, NodeId][] {
  const visited = new Set<NodeId>(); const disc = new Map<NodeId, number>(); const low = new Map<NodeId, number>();
  const parent = new Map<NodeId, NodeId | null>(); const result: [NodeId, NodeId][] = []; let timer = 0;
  const dfs = (u: NodeId): void => {
    visited.add(u); disc.set(u, timer); low.set(u, timer); timer++;
    for (const v of g.neighbors(u)) {
      if (!visited.has(v)) { parent.set(v, u); dfs(v); low.set(u, Math.min(low.get(u)!, low.get(v)!)); if (low.get(v)! > disc.get(u)!) result.push([u, v]); }
      else if (v !== parent.get(u)) low.set(u, Math.min(low.get(u)!, disc.get(v)!));
    }
  };
  for (const id of g.nodeIds()) if (!visited.has(id)) { parent.set(id, null); dfs(id); }
  return result;
}

export function isConnected(g: Graph): boolean { if (g.nodeCount() === 0) return true; const visited = new Set<NodeId>(); const stack = [g.nodeIds()[0]]; while (stack.length > 0) { const u = stack.pop()!; if (!visited.has(u)) { visited.add(u); for (const v of g.neighbors(u)) if (!visited.has(v)) stack.push(v); } } return visited.size === g.nodeCount(); }

export function isBipartite(g: Graph): { bipartite: boolean; partition?: [NodeId[], NodeId[]] } {
  const color = new Map<NodeId, 0 | 1>(); const p0: NodeId[] = []; const p1: NodeId[] = [];
  for (const start of g.nodeIds()) {
    if (color.has(start)) continue; color.set(start, 0); p0.push(start); const q = [start];
    while (q.length > 0) {
      const u = q.shift()!;
      for (const v of g.neighbors(u)) {
        if (!color.has(v)) { const c = (1 - color.get(u)!) as 0 | 1; color.set(v, c); (c === 0 ? p0 : p1).push(v); q.push(v); }
        else if (color.get(v) === color.get(u)) return { bipartite: false };
      }
    }
  }
  return { bipartite: true, partition: [p0, p1] };
}

export function biconnectedComponents(g: Graph): NodeId[][] {
  const visited = new Set<NodeId>(); const disc = new Map<NodeId, number>(); const low = new Map<NodeId, number>();
  const parent = new Map<NodeId, NodeId | null>(); const stack: [NodeId, NodeId][] = []; const comps: NodeId[][] = []; let timer = 0;
  const dfs = (u: NodeId): void => {
    visited.add(u); disc.set(u, timer); low.set(u, timer); timer++; let ch = 0;
    for (const v of g.neighbors(u)) {
      if (!visited.has(v)) {
        ch++; parent.set(v, u); stack.push([u, v]); dfs(v); low.set(u, Math.min(low.get(u)!, low.get(v)!));
        const isRoot = parent.get(u) === null || parent.get(u) === undefined;
        if ((isRoot && ch > 1) || (!isRoot && low.get(v)! >= disc.get(u)!)) {
          const comp = new Set<NodeId>();
          while (stack.length > 0) { const [a, b] = stack.pop()!; comp.add(a); comp.add(b); if (a === u && b === v) break; }
          comps.push([...comp]);
        }
      } else if (v !== parent.get(u) && disc.get(v)! < disc.get(u)!) { low.set(u, Math.min(low.get(u)!, disc.get(v)!)); stack.push([u, v]); }
    }
  };
  for (const id of g.nodeIds()) if (!visited.has(id)) { parent.set(id, null); dfs(id); }
  return comps;
}

File: generated_tasks/cryograph__22291410b572/environment/repo/src/components/components.cpp
#include "components/components.hpp"
#include <algorithm>
#include <queue>
#include <stack>
#include <stdexcept>

namespace cryo {

ComponentResult connected_components(const Graph& g) {
    ComponentResult res;
    std::unordered_set<NodeId> visited;
    auto ids = g.node_ids();

    for (auto start : ids) {
        if (visited.count(start)) continue;
        std::vector<NodeId> comp;
        std::queue<NodeId> q;
        q.push(start);
        visited.insert(start);

        while (!q.empty()) {
            NodeId cur = q.front();
            q.pop();
            comp.push_back(cur);
            for (auto& e : g.out_edges(cur)) {
                if (visited.insert(e.dst).second) q.push(e.dst);
            }
            if (g.is_directed()) {
                for (auto& [nid, edges] : std::unordered_map<NodeId, std::vector<Edge>>()) {}
                auto in = g.in_edges(cur);
                for (auto& e : in) {
                    if (visited.insert(e.src).second) q.push(e.src);
                }
            }
        }
        size_t idx = res.components.size();
        for (auto nid : comp) res.node_to_component[nid] = idx;
        std::sort(comp.begin(), comp.end());
        res.components.push_back(std::move(comp));
    }
    return res;
}

ComponentResult strongly_connected_components(const Graph& g) {
    ComponentResult res;
    auto ids = g.node_ids();
    size_t n = ids.size();
    if (n == 0) return res;

    std::unordered_map<NodeId, size_t> disc, low;
    std::unordered_map<NodeId, bool> on_stack;
    std::stack<NodeId> stk;
    size_t timer = 0;

    struct Frame {
        NodeId node;
        size_t edge_idx;
        bool returned;
    };

    for (auto nid : ids) {
        disc[nid] = SIZE_MAX;
        on_stack[nid] = false;
    }

    for (auto start : ids) {
        if (disc[start] != SIZE_MAX) continue;

        std::stack<Frame> call_stack;
        call_stack.push({start, 0, false});
        disc[start] = low[start] = timer++;
        stk.push(start);
        on_stack[start] = true;

        while (!call_stack.empty()) {
            auto& frame = call_stack.top();
            auto& edges = g.out_edges(frame.node);

            if (frame.returned) {
                NodeId child = edges[frame.edge_idx - 1].dst;
                low[frame.node] = std::min(low[frame.node], low[child]);
                frame.returned = false;
            }

            if (frame.edge_idx < edges.size()) {
                NodeId next = edges[frame.edge_idx].dst;
                frame.edge_idx++;

                if (disc[next] == SIZE_MAX) {
                    disc[next] = low[next] = timer++;
                    stk.push(next);
                    on_stack[next] = true;
                    call_stack.push({next, 0, false});
                } else if (on_stack[next]) {
                    low[frame.node] = std::min(low[frame.node], disc[next]);
                }
            } else {
                if (low[frame.node] == disc[frame.node]) {
                    std::vector<NodeId> comp;
                    while (true) {
                        NodeId w = stk.top();
                        stk.pop();
                        on_stack[w] = false;
                        comp.push_back(w);
                        if (w == frame.node) break;
                    }
                    size_t idx = res.components.size();
                    for (auto nid : comp) res.node_to_component[nid] = idx;
                    std::sort(comp.begin(), comp.end());
                    res.components.push_back(std::move(comp));
                }

                call_stack.pop();
                if (!call_stack.empty()) {
                    call_stack.top().returned = true;
                }
            }
        }
    }
    return res;
}

std::vector<std::pair<NodeId, NodeId>> find_bridges(const Graph& g) {
    std::vector<std::pair<NodeId, NodeId>> bridges;
    auto ids = g.node_ids();

    std::unordered_map<NodeId, size_t> disc, low;
    size_t timer = 0;

    struct Frame {
        NodeId node;
        NodeId parent;
        size_t edge_idx;
        bool returned;
    };

    for (auto nid : ids) disc[nid] = SIZE_MAX;

    for (auto start : ids) {
        if (disc[start] != SIZE_MAX) continue;

        std::stack<Frame> call_stack;
        disc[start] = low[start] = timer++;
        call_stack.push({start, INVALID_NODE, 0, false});

        while (!call_stack.empty()) {
            auto& frame = call_stack.top();
            auto& edges = g.out_edges(frame.node);

            if (frame.returned) {
                NodeId child = edges[frame.edge_idx - 1].dst;
                low[frame.node] = std::min(low[frame.node], low[child]);
                if (low[child] > disc[frame.node]) {
                    NodeId a = std::min(frame.node, child);
                    NodeId b = std::max(frame.node, child);
                    bridges.push_back({a, b});
                }
                frame.returned = false;
            }

            if (frame.edge_idx < edges.size()) {
                NodeId next = edges[frame.edge_idx].dst;
                frame.edge_idx++;

                if (disc[next] == SIZE_MAX) {
                    disc[next] = low[next] = timer++;
                    call_stack.push({next, frame.node, 0, false});
                } else if (next != frame.parent) {
                    low[frame.node] = std::min(low[frame.node], disc[next]);
                }
            } else {
                call_stack.pop();
                if (!call_stack.empty()) {
                    call_stack.top().returned = true;
                }
            }
        }
    }
    return bridges;
}

std::vector<NodeId> find_articulation_points(const Graph& g) {
    std::vector<NodeId> aps;
    auto ids = g.node_ids();
    std::unordered_map<NodeId, size_t> disc, low;
    std::unordered_set<NodeId> ap_set;
    size_t timer = 0;

    struct Frame {
        NodeId node;
        NodeId parent;
        size_t edge_idx;
        size_t children;
        bool returned;
    };

    for (auto nid : ids) disc[nid] = SIZE_MAX;

    for (auto start : ids) {
        if (disc[start] != SIZE_MAX) continue;

        std::stack<Frame> call_stack;
        disc[start] = low[start] = timer++;
        call_stack.push({start, INVALID_NODE, 0, 0, false});

        while (!call_stack.empty()) {
            auto& frame = call_stack.top();
            auto& edges = g.out_edges(frame.node);

            if (frame.returned) {
                NodeId child = edges[frame.edge_idx - 1].dst;
                low[frame.node] = std::min(low[frame.node], low[child]);
                frame.children++;

                if (frame.parent == INVALID_NODE && frame.children > 1) {
                    ap_set.insert(frame.node);
                }
                if (frame.parent != INVALID_NODE && low[child] >= disc[frame.node]) {
                    ap_set.insert(frame.node);
                }
                frame.returned = false;
            }

            if (frame.edge_idx < edges.size()) {
                NodeId next = edges[frame.edge_idx].dst;
                frame.edge_idx++;

                if (disc[next] == SIZE_MAX) {
                    disc[next] = low[next] = timer++;
                    call_stack.push({next, frame.node, 0, 0, false});
                } else if (next != frame.parent) {
                    low[frame.node] = std::min(low[frame.node], disc[next]);
                }
            } else {
                call_stack.pop();
                if (!call_stack.empty()) {
                    call_stack.top().returned = true;
                }
            }
        }
    }

    aps.assign(ap_set.begin(), ap_set.end());
    std::sort(aps.begin(), aps.end());
    return aps;
}

bool is_bipartite(const Graph& g) {
    std::unordered_map<NodeId, int> coloring;
    return is_bipartite(g, coloring);
}

bool is_bipartite(const Graph& g, std::unordered_map<NodeId, int>& coloring) {
    auto ids = g.node_ids();
    for (auto nid : ids) coloring[nid] = -1;

    for (auto start : ids) {
        if (coloring[start] != -1) continue;
        std::queue<NodeId> q;
        q.push(start);
        coloring[start] = 0;

        while (!q.empty()) {
            NodeId cur = q.front();
            q.pop();
            for (auto& e : g.out_edges(cur)) {
                if (coloring[e.dst] == -1) {
                    coloring[e.dst] = 1 - coloring[cur];
                    q.push(e.dst);
                } else if (coloring[e.dst] == coloring[cur]) {
                    return false;
                }
            }
        }
    }
    return true;
}

bool is_connected(const Graph& g) {
    if (g.node_count() <= 1) return true;
    auto ids = g.node_ids();
    std::unordered_set<NodeId> visited;
    std::queue<NodeId> q;
    q.push(ids[0]);
    visited.insert(ids[0]);

    while (!q.empty()) {
        NodeId cur = q.front();
        q.pop();
        for (auto& e : g.out_edges(cur)) {
            if (visited.insert(e.dst).second) q.push(e.dst);
        }
    }
    return visited.size() == g.node_count();
}

bool is_strongly_connected(const Graph& g) {
    if (g.node_count() <= 1) return true;
    auto scc = strongly_connected_components(g);
    return scc.count() == 1;
}

Graph condensation(const Graph& g) {
    auto scc = strongly_connected_components(g);
    Graph cond(GraphType::Directed);
    for (size_t i = 0; i < scc.count(); i++) {
        cond.add_node_with_id(i);
    }
    std::set<std::pair<size_t, size_t>> added;
    g.for_each_edge([&](const Edge& e) {
        size_t cs = scc.node_to_component.at(e.src);
        size_t cd = scc.node_to_component.at(e.dst);
        if (cs != cd && added.insert({cs, cd}).second) {
            cond.add_edge(cs, cd);
        }
    });
    return cond;
}

size_t largest_component_size(const Graph& g) {
    auto cc = connected_components(g);
    size_t max_size = 0;
    for (auto& c : cc.components) max_size = std::max(max_size, c.size());
    return max_size;
}

size_t component_count(const Graph& g) {
    return connected_components(g).count();
}

} // namespace cryo